### PR TITLE
igraph_is_graphical and igraph_is_bigraphical

### LINKED ISSUE #1432: Graphicality for non-simple graphs
`igraph_is_graphical_degree_sequence` checks if a degree sequence has a realization as a *simple* graph and `igraph_realize_degree_sequence` constructs it.

I am going to implement checking if there is a realization as a graph with/without self-loops and with/without multi-edges, and eventually also constructing these realizations (because I need it for my work).

I am looking for feedback on the API. We can:

 1. create new functions
 2. extend the existing `igraph_is_graphical_degree_sequence`

I prefer (2), as we can use a consistent extension to both `is_graphical_degree_sequence` and `realize_degree_sequence`.

If we go with (2), my first idea was to add two arguments, `igraph_bool_t multiple, igraph_bool_t loops`, similar to other existing functions. These control if multi-edges and self-loops are allowed respectively. This is not quite ideal because `multiple=false, loops=true` would naturally implies that while self-loops may be present, there may only be at most one. Implementing this case is straightforward enough, but it's kind of pointless (I do not expect many applications). The interesting cases are _loopless multigraphs_ and _loopy multigraphs_. Thus, the next possibility would be to have a single "type" variable with enum values such as `SIMPLE_GRAPH`, `LOOPLESS_MULTIGRAPH`, `LOOPY_MULTIGRAPH`. Not as pretty, but probably more aligned with actual use cases, and saves us implementing some not-so-useful cases. It would also allow specifying that we allow multiple self-loops, but we do not allow multiple edges between distinct vertices (which is probably not needed, but it is considered in papers occasionally). Yet another possibility is to go with the `multiple` and `loops` flags but simply return an `IGRAPH_UNIMPLEMENTED` for the missing cases.

So, the choices are:

 1. `multiple` and `loops` flags, complete implementation right away
 2. `multiple` and `loops` flags, partial implementation for now, return `IGRAPH_UNIMPLEMENTED` for the `multiple=false` and `loops=true` case.
 3. Single control variable with named options.

Do you have any preferences on which one to go with?

----

Note 1: We do already have `igraph_is_degree_sequence`, which is basically the loopy multigraph case. 

Note 2: Eventually, a similar extension could be implemented for `igraph_rewire` as well, but this is more work (especially the directed case).

### PR DESCRIPTION
**Pull request by szhorvat**
Closes #1432 .

This draft is open for general comments on the API/design/documentation. Feedback is welcome.

I still need to verify the implementation, add tests, and possibly complete `igraph_i_is_graphical_undirected_loopy_simple` if all goes well. 

I chose the `SW` suffix constants that represent bitwise switches and can be combined with `|`.

### CODE REVIEW

**Comment by szhorvat** [comment]
The intent is to deprecate `igraph_is_degree_sequence` and `igraph_is_graphical_degree_sequence` in favour of `igraph_is_graphical`.  I opted to leave the implementations of `igraph_is_degree_sequence` and `igraph_is_graphical_degree_sequence` as they are (even though a bit of code is duplicated) because there are some tiny differences: if the in- and out-degree sequence were not of the same length, previously `false` was returned. Now it throws an error for this case.

**Comment by szhorvat** [comment]
I should note that I plan to extend `igraph_realize_degree_sequence` in the same way, but that will be a followup PR.

**Comment by szhorvat** [comment]
@ntamas No objections so far with the bit-switches (even though one of the constants I introduced set 2 bits at once)? :-)

**Comment by ntamas** [comment]
Seems to be okay so far! I haven't checked the actual implementation steps yet, only the API. I'll go through the implementation once I'm back from holidays (Thursday-ish).

**Comment by szhorvat** [comment]
> I haven't checked the actual implementation steps yet,

Don't do that yet, I am still making changes. I only wanted feedback on the API. Thanks!

**Comment by szhorvat** [comment]
I gave up hope on keeping commits organized. We are going to do squash-and-merge. Since the last two commits are all mixed up and only make sense together, please only look at the final result when reviewing.

**Comment by ntamas** [comment]
Will do, thanks! (That's my default mode of operation anyway).

**Comment by szhorvat** [comment]
A summary of what's in this PR:

 - Deprecate `is_graphical_degree_sequence` and `is_degree_sequence` in favour of `is_graphical`.
 - `is_graphical` now works for directed/undirected graphs, loops/no-loops, multi-edges/simple-edges. That's `2*2*2 = 8` cases.
 - The old `is_degree_sequence` corresponded to loops/multi-edges.
 - Undirected simple graph case gets a new implementation (Cloteaux), which is both faster and O(n) (not O(n log n)).
 - Undirected loopy graph case uses a modified version of the previous simple graph case implementation (Király). It could be sped up with a similar approach to the simple graph case, but I simply don't have the time to do this now. 
 - Realizing degree sequences is now possible also for loopless and loopy undirected multigraphs, with a nice analogue of the three Havel-Hakimi variants we used for the simple graph case. It even works out that `IGRAPH_REALIZE_DEGSEQ_SMALLEST` guarantees that we get a connected graph. *Why* this works is not entirely trivial, if you want the proof just email me.
 - I did not implement graph realization for the loopy simple graph case yet. There is a simple way to do it (make a loop on every vertex with degree >=2 before we start), but it's not a nice analogue of the methods used for the other cases. I want something nicer, but don't have the time to fuss with the math right now.  Same goes for directed cases. The directed case is easier and I'll probably do it later this summer, but not in this PR. If everything were implemented, we'd have `2*2*2*3 = 24` cases (!) with the three kinds of methods ...
 - Some bugs were fixed in realizing degree sequences: sometimes negative degrees were not detected.
 - There's some ugliness in how some tests are done. I used macros instead of functions. Functions would be better, but macros work, and I don't have the energy to make this prettier ...

**Comment by szhorvat** [comment]
@ntamas A careful review would be appreciated, but please don't merge it, just let me know if you find any issues, or if anything is unclear in the docs. This was a lot more work than I expected ...

I'll have another read-through in a couple of days and I'll merge it myself (after you approved it).

**Comment by szhorvat** [comment]
@ntamas I added some more tests today which revealed some (old) issues. I will commit the fix to those directly to `master` as well.  It's good now, ready for review. AppVeyor fails again, but that's not because of this PR.

**Comment by ntamas** [comment]
AppVeyor failures are caused by [this issue](https://help.appveyor.com/discussions/problems/27791-msys2-keyring-on-vs2017-image-at-least-needs-update). I'd rather wait a few more days and see if they fix it on AppVeyor's side rather than introducing a fix on our side (i.e. importing the missing PGP key manually) because it is likely to affect all projects using MSYS2 so chances are that something will be done in AppVeyor anyway.

**Comment by ntamas** [comment]
Also, I'll start reviewing it tomorrow, thanks for your work on this!

**Comment by szhorvat** [comment]
What I like to do when reviewing new functionality is to integrate it into the Mathematica interface.

It would be really nice if this (especially [realize_degree_sequence](https://github.com/igraph/python-igraph/issues/316)) could be integrated into the Python interface. Then I could also reference it in my paper, where I show that the "smallest" method always creates a connected graph.

BTW there's no hurry at all in reviewing this. It is not blocking any of my current projects.

**Comment by ntamas** [comment]
Seems okay to me in general - only a few minor changes requested.

File: src/graphicality.c [hunk 1]
```diff
@@ -0,0 +1,936 @@
...
+ *
+ * </para><para>
+ * In particular, the function checks whether all the degrees are non-negative.
+ * For undirected graphs, it also checks whether the sum of degrees is even.
+ * For directed graphs, the function checks whether the lengths of the two
+ * degree vectors are equal and whether their sums are also equal. These are
+ * known sufficient and necessary conditions for a degree sequence to be
+ * valid.
+ *
+ * \deprecated-by igraph_is_graphical 0.9
```
[the remark this answers is not in the document]
Comment (szhorvat) [reply src/graphicality.c:852]: Absolutely! But I noticed that you placed it last in other functions (unless I misread it). That's why I did this. I'll move it to the top.

File: src/graphicality.c [hunk 2]
```diff
@@ -0,0 +1,936 @@
...
+/**
+ * \function igraph_is_graphical
+ * \brief Is there a graph with the given degree sequence?
+ *
+ * Determines whether a sequence of integers can be the degree sequence of some graph.
+ * The classical concept of graphicality assumes simple graphs. This function can perform
+ * the check also when either self-loops, multi-edge, or both are allowed in the graph.
+ *
+ * </para><para>
+ * For simple undirected graphs, the Erdős-Gallai conditions are checked using the linear-time
```
[the remark this answers is not in the document]
Comment (szhorvat) [reply src/graphicality.c:50]: It passes, but I didn't check if UTF-8 is *actually supported* or just happens to work accidentally ...

**Comment by szhorvat** [comment]
@ntamas One more thing: When I fixed the Erdos-Gallai test on `master`, I added the `igraph_vector_reverse_sort` function, and also added it to the documentation. If there is no objection, I am going to remove it from the docs on `master` and keep it on `develop` only. There's some way til 0.9 and 0.8.3 is not the right time to officially introduce new functions. I'll keep the function (since it's actually needed for the fix), but I won't include it in the docs.

File: src/graphicality.c [hunk 3]
```diff
@@ -0,0 +1,936 @@
...
+    long n = igraph_vector_size(degrees);
+
+    for (i=0; i < n; ++i) {
+        long int d = VECTOR(*degrees)[i];
+
+        if (d < 0) {
+            *res = 0;
+            return IGRAPH_SUCCESS;
+        }
+        sum += d;
```
[the remark this answers is not in the document]
Comment (szhorvat) [reply src/graphicality.c]: Fixed. I opted to fix it only here, but not in the Erdos-Gallai test as there we check that `dmax <= n-1`, so problems from overflow are exceedingly unlikely.

Actually, if I think about it: parity should be maintained even if there is integer overflow.

**Comment by szhorvat** [comment]
I addressed everything you pointed out. I'll merge this myself in a couple of days.

**Pull request merged**